### PR TITLE
Add batching rule for torch.sum(tensor, dims)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39859 Batching rule for torch.mul
* **#39581 Add batching rule for torch.sum(tensor, dims)**
* #39580 Impose maximum level restriction for BatchedTensors

Context: Batching rules
------------------------------------

Batching rules take BatchedTensors and regular Tensors as arguments. A
batching rule generally does the following:

1. Converts (logical) BatchedTensors to views on physical tensors.
2. Converts logical arguments (e.g. dimension indexes, shapes) to
physical arguments that correspond to the physical tensors.
3. Calls at:: operations on the physical tensors and arguments.
4. Converts physical results back to BatchedTensors.

Steps 1 and 2 differ for operators with different batching behaviors.
(see next section)

VmapTransform abstraction
------------------------------------
(Previously known as a "Converter". Bikeshedding welcome on the naming).

An ArgTransform converts logical views of tensors to physical views.
When
writing a batching rule, users should select the ArgTransform that
matches
the batching behavior of their operator. If the batching behavior of the
op is complicated, then they’ll have to write some custom logic (either
by writing a new ArgTransform, or writing the logical->physical
transform
themselves).

*56% (~474) of (vmap-supported) operators can and will use these
VmapTransform. 20% (~168) of operators need custom handling*.

See `VmapTransforms.h` for more context.

PhysicalView
------------------------------------

VmapTransforms return physical views on tensors, represented by the
PhysicalView struct. It is effectively a Tensor and contains
enough metadata to enable mapping logical non-tensor arguments to
physical non-tensor arguments, and the other way around.

There are two methods on PhysicalView right now:
- `PhysicalView::getPhysicalDim(logical_dim)` and
`PhysicalView::getPhysicalDims(logical_dims)`.
are used to map logical dims to physical dims.
- `PhysicalView::newLogicalFromPhysical(Tensor)` is used to map a result
physical tensor from a batching rule to a logical tensor
(BatchedTensor).

Test Plan:
------------------------------------
- `./build/bin/vmap_test`

Differential Revision: [D21983789](https://our.internmc.facebook.com/intern/diff/D21983789)

### CODE REVIEW

**Comment by ezyang** [comment]
> (Previously known as a "Converter". Bikeshedding welcome on the naming).

Can we explicitly put "Batch" or "VMap" in the name? "VMapTransform"?

File: aten/src/ATen/BatchingArgTransforms.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+struct TORCH_API PhysicalView {
+  PhysicalView(Tensor&& tensor)
+      : tensor_(tensor) {
+    TORCH_INTERNAL_ASSERT(!isBatched(tensor));
+  }
+
+  Tensor& tensor() { return tensor_; }
+
+ private:
+  Tensor tensor_;
```
Comment (ezyang) [inline aten/src/ATen/BatchingArgTransforms.h:64]: This doesn't feel like sufficient metadata in PhysicalView?

**Comment by ezyang** [comment]
This might be a case where it's easier to review the stack of PRs in one go, rather than one by one (fyi @cpuhrsch  too)

**Comment by zou3519** [comment]
@ezyang @cpuhrsch I tagged you as reviewers on the full stack. Would it be easier to review if I just squashed all of the commits together into one single commit?

File: aten/src/ATen/BatchingArgTransforms.h (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingArgTransforms.h:64]: It's implemented in the next PR up: https://github.com/pytorch/pytorch/pull/39593

**Comment by ezyang** [comment]
Yes, I think squashing would be easier, especially since some later PRs modify code in earlier PRs.

**Comment by zou3519** [comment]
@ezyang sounds good, let me squash and re-push

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+// Note: [Future plans]
+// The API for writing a batching rule isn't stable. In the future, we'd like
+// to think about the problem of translating these batching rules to TorchScript.
+// Ideally batching rules in eager mode vs TorchScript would look pretty similar,
+// if not use the same mechanism. In order to accomplish that we might have to
+// do some refactoring.
+
+Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, optional<ScalarType> dtype) {
+  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
```
Comment (ezyang) [inline aten/src/ATen/BatchingRegistrations.cpp:49]: Reading over this, it also occurs to me that "logical" and "physical" are also very generic concepts which could apply to a variety of situations, not just vmap. So maybe we should try to find some sort of vmap-y nomenclature here. A really silly name is "batched" and "unbatched"; the "batched" naming makes sense, but unbatched doesn't really make any semantic sense, except maybe in the sense that unbatched tensors are "not" batch tensors (but they're still batched, just implicitly). After spending another thirty seconds thinking about it, I couldn't think of anything good, so maybe logical-physical is fine (since these seem to always be contextualized on one of these vmap transform things)

File: aten/src/ATen/VmapTransforms.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+// [2, 3, 4].
+//
+// This notion of logical vs physical also extends to non-tensor arguments.
+// Consider the previous tensor; let's assume the user called
+// `torch.sum(tensor, dim=0)` inside of `func`. Then the logical
+// dimension they are reducing over is dim 0 but the physical dim is dim 1
+// (the first non-batch dimension)
+
+// Forward declared; see NOTE: [What is a PhysicalView?]
+struct PhysicalView;
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.h:29]: But this struct name isn't good, because it doesn't reveal that it has anything to do with batching. (Hmm... maybe UnbatchedView isn't such a bad name after all...)

File: aten/src/ATen/VmapTransforms.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+//
+// One can use it to further convert logical dimension indices, logical shapes,
+// and more to their physical variants, or convert a new (physical) tensor into
+// a logical BatchedTensor. (TODO(rzou): some of these are not yet implemented).
+//
+// PhysicalView stores a physical tensor with all of its batch dimensions at
+// the front and some levels that correspond to said batch dimensions.
+//
+// The levels contain the information necessary to perform the logical<->physical
+// mapping. For example, given:
```
Comment (ezyang) [inline aten/src/ATen/VmapTransforms.h:60]: This is a really indirect way to describe what "levels" is. Don't say what the information is used to do: tell me what the information means. "The levels bitset specifies which vmap levels correspond to the batch dimensions at the front of the tensor. In particular, the number of set bits corresponds to the number of batch dimensions on tensor, and the rightmost bit of levels specifies the minimum number of nested vmaps we are in at this point of time."

Also, this particular representation of PhysicalView doesn't seem to generalize to other vmap transforms. For example, if you have a tensor that accepts only one batch dimension, you're going to need to store something else in this case, right?

**Comment by ezyang** [comment]
Broadly speaking this looks pretty good, and I'd be happy to land something and start iterating on it. Would like it if @cpuhrsch takes a look at it though, since I am a bit biased (having been in the original conversations about this).

File: aten/src/ATen/VmapTransforms.h (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/VmapTransforms.h:60]: > Also, this particular representation of PhysicalView doesn't seem to generalize to other vmap transforms. For example, if you have a tensor that accepts only one batch dimension, you're going to need to store something else in this case, right?

Yeah, you'd have to store levels and the batch sizes. I'm not sure if I should reuse the PhysicalView class when that happens (and add an optional<batch_sizes> field), or just make a different class for it and call it "FlattenedBatchPhysicalView"

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:49]: I think we've overloaded the word "batch" here and that is why "batched" and "unbatched" seem weird. What we mean by a "batchedtensor" is a "vmap'ed tensor" and a "batch dim" iss a "dimension being vmap'ed over", so maybe these things should be called "VmappedTensor" and "Vmapped dim" instead and we should use "Vmaped" view and "UnVmaped" view instead of "logical" and "physical"...

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+// When you are adding a new operator, you'll need to add a batching rule so
+// that vmap can work efficiently with said operator. If you do not, we'll attempt
+// to generate a slow fallback for the batching rule (this is not yet implemented).
+
+// NOTE: [How to write batching rules?]
+// The signature of a batching rule should look like exactly like the C++ signature
+// of its operator.
+//
+// At a high level, what a batching rule does is the following:
+// 1. Converts (logical) BatchedTensors to views on physical tensors.
```
Comment (cpuhrsch) [inline aten/src/ATen/BatchingRegistrations.cpp:28]: I think this could be easier to read for novices if you define what you mean by logical and physical first

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+// 2. Converts logical arguments (e.g. dimension indexes, shapes) to physical
+//    arguments that correspond to the physical tensors.
+// 3. Calls at:: operations on the physical tensors and arguments to produce
+//    some physical results.
+// 4. Converts physical results back to BatchedTensors.
+//
+// Steps 1, 2, and 4 differ for operators with different batching behaviors. When
+// writing a new batching rule, please select an VmapTransform that matches the
```
Comment (cpuhrsch) [inline aten/src/ATen/BatchingRegistrations.cpp:36]: nit: "a" instead of "an"

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (cpuhrsch) [reply aten/src/ATen/BatchingRegistrations.cpp:49]: What about "representation" and "implementation"? A BatchedTensor is a presentation for a Tensor with batch dimensions, which represents a collection of Tensors of same size. The Tensor underlying a BatchedTensor is the implementation. It implements this collection of Tensors by packing them into a single Tensor of higher dimension, where some dimensions are reserved to implement batching only and not exposed to the user. It's these dimensions that we need to navigate when implementing a kernel.

File: aten/src/ATen/VmapTransforms.h (hunk 4 above)
Comment (cpuhrsch) [reply aten/src/ATen/VmapTransforms.h:60]: How about: PhysicalView ties the packed data Tensor underlying a BatchedTensor with a bitset annotates the dimensions of the packed data Tensor as either user visible (if the corresponding bit is set to 1) or implementation-only visible (if the corresponding bit is set to 0).

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 5 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:28]: I'll link to the physical vs logical definition on the other source file. I agree that the current organization is not the best; ideally we'd have a "batching" or "vmap" namespace and folder that contains all of the vmap implementation details with a README that details what is going on and the right order to read the code.

**Review by ezyang — APPROVED** [review]
